src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,144 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            head = new Node<T>(null, value, null);
+            tail = head;
+        } else {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("The index "
+                    + index + " passed to method is invalid");
+        }
+        if (size == index) {
+            add(value);
+        } else {
+            Node<T> currentNode = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev = newNode;
+            if (index == 0) {
+                head = newNode;
+            } else {
+                newNode.prev.next = newNode;
+            }
+            size++;
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexValidation(index);
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexValidation(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.item;
+        currentNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexValidation(index);
+        Node<T> node = findNodeByIndex(index);
+        if (head == tail) {
+            head = null;
+            tail = null;
+        } else if (index == 0) {
+            head.next.prev = null;
+            head = head.next;
+        } else if (index == size - 1) {
+            tail.prev.next = null;
+            tail = tail.prev;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
```

Comment:
let's move this logic to the separate method unlink() or removeNode() and call it here